### ISSUE TITLE
`--watch` Only watches files below `process.cwd()`, even though some paths may be outside of it

### ISSUE THREAD

**Issue by milesrichardson** [dotansimha/graphql-code-generator#9266]
### Which packages are impacted by your issue?

@graphql-codegen/cli

### Describe the bug

Watch mode only watches files below `process.cwd()`, even though the codegen config may validly reference files outside of this directory. 

**Bug**

This bug was introduced in #9009. Previously, when using Chokidar, watch mode listened for changes in every relevant file. However, after switching to `@parcel/watcher`, [watch mode only listens for changes below `process.cwd()`](https://github.com/dotansimha/graphql-code-generator/blob/master/packages/graphql-codegen-cli/src/utils/watcher.ts#L95-L97), and then filters each change to see whether it's a relevant file. This is a fine approach, except that it's possible and valid for a config to reference paths outside of the current directory. When calling `graphql-codegen generate`, the bug does not exist, and files outside of `process.cwd()` are included in the generation. Therefore, this is a regression in watch mode only.

**Reproduction**

I have a minimal reproduction, but unfortunately watch mode doesn't work in Stackblitz, so you need to pull the repo and run it locally to see it reproduce:

- [Repository with reproduction](https://www.github.com/milesrichardson/repro-npm-graphql-codegen-watch-outside-cwd)
- [Stackblitz with reproduction](https://stackblitz.com/github/milesrichardson/repro-npm-graphql-codegen-watch-outside-cwd/?file=README.md) (cloned from the above repo)

**Fix**

The fix, if we want to match the previous behavior, is to find the "longest common prefix" directory of all relevant files, and watch that directory instead of `process.cwd()`. An alternative solution would be to allow the user to specify a `config.watchDir`, but this would still be a regression from previous behavior. 

I have a PR which implements the "longest common prefix." I will open it shortly, after I submit this issue.

EDIT: A pull request is pending review: #9267 

### Your Example Website or App

https://www.github.com/milesrichardson/repro-npm-graphql-codegen-watch-outside-cwd

### Steps to Reproduce the Bug or Issue

See readme of reproduction repo.

### Expected behavior

As a user, I expect that if I can reference a path outside of the current directory (e.g. `documents: "../somewhere-else/*.graphql"`), and if this correctly generates output when calling `graphql-codegen`, that calling `graphql-codegen --watch` should also include that path in watch mode. However, the bug is that only files below the current working directory are included in watch mode, even though they are correctly included in build mode.

### Platform

- OS: all
- NodeJS: all
- `graphql` version: irrelevant
- `@graphql-codegen/cli` version(s): 3.2.2

**Comment by saihaj** [comment]
can you test [`@graphql-codegen/cli@3.3.1-rc-20230404030530-47bcfb095`](https://github.com/dotansimha/graphql-code-generator/pull/9269#issuecomment-1495279376)

**Comment by milesrichardson** [comment]
sure thing, will test it in the morning

**Comment by milesrichardson** [comment]
I tested this in our repository, and it works as expected. I think you can cut that release.

---

However, in our case, this exposed a different bug, which sounds like what I've seen complaints about in a few other issues. Basically, the [event filtering is using `.contains()`](https://github.com/dotansimha/graphql-code-generator/blob/master/packages/graphql-codegen-cli/src/utils/watcher.ts#L107), but I believe it [should be using `.isMatch()`](https://github.com/micromatch/micromatch#ismatch), because [`.contains()` includes partial matches](https://github.com/micromatch/micromatch#contains). In our case this meant that when a rebuild caused changes in the `.next` directory, _those changes_ were also included because they were a "partial match", e.g.: 

```
Running lifecycle hook "onWatchTriggered" scripts...
[Watcher] triggered due to a file create event: /src/js/cloud/catalog/.next/cache/webpack/server-development/7.pack
```

I patched our local installation to switch to `.isMatch` and it seems to work as expected. I'll open an issue/PR for that bug/fix as well, but I'm slightly less confident in it because `micromatch` rules are notoriously finnicky and I don't know how everyone else is using them.

**Comment by milesrichardson** [comment]
Added some tests for this behavior (and fixed a few minor bugs) as part of #9275
